### PR TITLE
Refactor to allow more customization via hooks and how Qt UI widgets are built

### PR DESCRIPTION
**Pull request by staceyoue**
This refactor is to allow `tk-vred` to customize SG Panel when launching VRED Presenter to review a list of Versions from a Playlist (e.g. launched via Toolkit Action on Playlist entity).

This change helps resolve these JIRA tickets:

- SHOT-3555
- SHOT-3556
- SHOT-3608
- SHOT-3634
- SHOT-3633

Major changes:

- Entity tabs are created at runtime now, instead of defined in `.ui` file
- Entity tabs are built based on settings defined in hooks
- All entity tabs are handled the same (e.g. no special case handling for Version, PublishedFile)
- New shotgun_filters_hook

### CODE REVIEW

**Comment by thebeeland** [comment]
Hey @staceyoue -- I'm going to work my way through this PR next. It's pretty big, but your explanation in the PR description makes sense, so I think I have what I need to get started. It might take me a few days to look it all over, but hopefully by next week sometime I'll have submitted a full review.

**Comment by thebeeland** [comment]
Actually, one thing @staceyoue -- there's a conflict in dialog.py that'll need to be resolved. Mind going ahead and doing that so I can see the final result while I'm reviewing everything else? I've not looked at what the conflict is, but hopefully it's something minor.

File: python/app/dialog.py
Comment (staceyoue) [reply python/app/dialog.py:533]: Hm, running the pre-commit hook locally does not pick up this issue..

**Comment by staceyoue** [comment]
> Actually, one thing @staceyoue -- there's a conflict in dialog.py that'll need to be resolved. Mind going ahead and doing that so I can see the final result while I'm reviewing everything else? I've not looked at what the conflict is, but hopefully it's something minor.

Resolved!

As I reviewed my own PR after having not looked at it in quite some time, I was thinking that I should improve how the entity tab data is created for the panel (in `build_entity_tabs`). Mainly that the tab data is managed using a dictionary, which is getting quite complex -- my thoughts now are to create a separate class to handle that data.

I'll let you take a look first though and see what you think, before I go changing anything.

**Comment by staceyoue** [comment]
@thebeeland also for your reference, here are the dependent/related PRs:

**tk-framework-desktopserver**:  this is to enable the Playlist entity for toolkit actions
https://github.com/shotgunsoftware/tk-framework-desktopserver/pull/77

**tk-config-default2**/**tk-config-basic**:  configs to enable Playlist toolkit action to launch "Review with VRED"
https://github.com/shotgunsoftware/tk-config-default2/pull/96/files
https://github.com/shotgunsoftware/tk-config-basic/pull/101

**tk-shotgun-launchvredreview**:  allow launching VRED with Playlist entity
https://github.com/shotgunsoftware/tk-shotgun-launchvredreview/pull/3

**tk-vred**: added hooks to override the SG panel default hooks to create Playlist tab (shotgun_fields.py, shotugn_filters.py)
https://github.com/shotgunsoftware/tk-vred/pull/51

File: hooks/shotgun_filters.py
Comment (thebeeland) [inline hooks/shotgun_filters.py:101]: What's this mean? It's confusing when taking 102-104 into consideration.

File: hooks/shotgun_filters.py
Comment (thebeeland) [inline hooks/shotgun_filters.py:23]: I'm guessing this might be old logic from Manne that's being pushed into a hook now. If that's true, then apologies for asking you to clean up a couple things below.

File: hooks/shotgun_filters.py
Comment (thebeeland) [inline hooks/shotgun_filters.py:220]:
```suggestion
        self.logger.debug(
            "%s Resolved %s into the following sg query:\n%s",
            (self, sg_location, pprint.pformat(link_filters))
        )
```

File: python/app/dialog.py
Comment (thebeeland) [inline python/app/dialog.py:321]: Mind doing this from a try/finally to make sure it's set back to `False` even if something goes wrong in between?

File: python/app/dialog.py
Comment (thebeeland) [reply python/app/dialog.py:533]: Don't sweat this particular message when you get it from Hound.

File: python/app/model_entity_listing.py
Comment (thebeeland) [inline python/app/model_entity_listing.py:113]: Nitpick: `TypeError`

**Review by thebeeland — CHANGES_REQUESTED** [review]
Only a few small things. This was a lot of work, but I see you've gone to some lengths to ensure that out of the box it remains the same as before. We'll want to confirm that's true before merging, but I'm guessing it'll look good based on what I see here.

File: hooks/general_actions.py
Comment (staceyoue) [reply hooks/general_actions.py:191]: Yes, our VRED integration returns a value. I'll update this here to `return dict()`

File: hooks/general_actions.py
Comment (staceyoue) [reply hooks/general_actions.py:250]: You're right, there is no need to return this value since the caller has it already. I introduced this hook so that our VRED integration can do some other actions on double click action, but there is no need to return the entity that is being passed in -- I'll fix this method to have no return value

Comment (staceyoue) [reply hooks/general_actions.py:250]: I was also a little confused why I had this returning as well after looking at it again. So the reason I have the entity dict returning is basically to signal to the caller (the dialog) to proceed or cancel the double click action.

In our VRED integration, there are cases where we want to abort the double-click to navigate to the entity in the panel.

I think it would be better to then just return True or False? And the caller has the entity data to navigate, so no need to return that data back.

File: hooks/shotgun_fields.py
Comment (staceyoue) [reply hooks/shotgun_fields.py:436]: > what does it mean for the activity stream tab to have that notes sub-dict enabled vs. disabled?

The `notes` is actually its own key-value pair (not a sub-dict of the `activity` stream dict value) -- the formatting I think just makes it look like it's part of the activity dict

File: hooks/shotgun_filters.py
Comment (staceyoue) [reply hooks/shotgun_filters.py:23]: Yeah, I copied this `get_link_filters` method from the dialog to expose it in the hook. No problem though, happy to improve things while we're here anyway.

File: hooks/shotgun_filters.py
Comment (staceyoue) [reply hooks/shotgun_filters.py:101]: Not sure to be honest. I'll update this comment

File: python/app/dialog.py
Comment (staceyoue) [reply python/app/dialog.py:321]: Definitely

File: hooks/shotgun_filters.py
Comment (staceyoue) [reply hooks/shotgun_filters.py]: This suggested change is causing an error on Windows:
```
TypeError: not enough arguments for format string
```

So I've left it as is for now.

**Comment by staceyoue** [comment]
@thebeeland confirmed that the default display and behaviour has not changed with these code changes

File: hooks/general_actions.py
Comment (thebeeland) [reply hooks/general_actions.py]: I think True/False for whether to proceed with the double-click action makes sense to me. 👍

File: hooks/shotgun_filters.py
Comment (thebeeland) [reply hooks/shotgun_filters.py]: Okay. That's weird, but of little/no consequence. Let's just leave it as is.

**Review by thebeeland — APPROVED** [review]
Based on your responses to the remaining comments, I'm going to go ahead and approve this. Once you're happy with everything you're free to carry on!

**Pull request merged**
